### ISSUE TITLE
[策略3] 机器学习趋势预测 | ML Trend Prediction

### ISSUE THREAD

**Issue by Kaoruha** [Kaoruha/GinkGO#4421]
**标签**: quantitative-strategy, machine-learning, lstm, trend-prediction
### 摘要
使用LSTM和注意力机制预测价格趋势，生成动态交易信号。

### 算法描述
**核心原理：**
LSTM（长短期记忆网络）能有效捕捉时间序列中的长期依赖关系，结合注意力机制可动态调整不同时间步的权重，使模型聚焦于最关键的预测信号。

**核心公式：**
1. **LSTM单元计算：**
   ```
   f_t = σ(W_f * [h_{t-1}, x_t] + b_f)  # 遗忘门
   i_t = σ(W_i * [h_{t-1}, x_t] + b_i)  # 输入门
   C_t = f_t * C_{t-1} + i_t * tanh(W_c * [h_{t-1}, x_t] + b_c)  # 细胞状态
   o_t = σ(W_o * [h_{t-1}, x_t] + b_o)  # 输出门
   h_t = o_t * tanh(C_t)  # 隐藏状态
   ```

2. **注意力机制：**
   ```
   e_t = v^T * tanh(W_a * h_t)
   α_t = softmax(e_t) = exp(e_t) / Σ exp(e_s)
   context = Σ α_t * h_t
   ```

3. **特征构建：**
   ```
   X_t = [return_t-n, ..., return_t-1, return_t, volume_t, volatility_t, sma_t]
   其中 return_t = (close_t - close_{t-1}) / close_{t-1}
   ```

4. **信号生成：**
   ```
   P(trend_up) = sigmoid(W_out * [h_last, context] + b_out)
   if P > 0.6: BUY_SIGNAL
   if P < 0.4: SELL_SIGNAL
   else: HOLD
   ```

**模型架构：**
- Input: 60历史数据点，5个特征
- Layer1: LSTM(64 units, return_sequences=True)
- Layer2: Attention Layer
- Layer3: LSTM(32 units)
- Layer4: Dense(16, ReLU)
- Output: Dense(1, Sigmoid)

### Python代码示例
```python
import numpy as np
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from dataclasses import dataclass
from typing import List, Tuple, Optional
import warnings
warnings.filterwarnings('ignore')

@dataclass
class MarketFeature:
    returns: np.ndarray
    volume: np.ndarray
    volatility: np.ndarray
    sma_ratio: np.ndarray

@dataclass
class TrendSignal:
    direction: str
    confidence: float
    predicted_return: float
    holding_period: int

class MLTrendPredictor:
    def __init__(self, lookback: int = 60, prediction_horizon: int = 5):
        self.lookback = lookback
        self.prediction_horizon = prediction_horizon
        self.scaler = StandardScaler()
        self.weights = None
        self.bias = None
        
    def compute_features(self, data: pd.DataFrame) -> MarketFeature:
        close = data['close'].values
        volume = data['volume'].values
        
        returns = np.diff(close) / close[:-1]
        returns = np.concatenate([[0], returns])
        
        volatility = pd.Series(close).rolling(20).std().values
        sma_20 = pd.Series(close).rolling(20).mean().values
        sma_50 = pd.Series(close).rolling(50).mean().values
        sma_ratio = sma_20 / (sma_50 + 1e-10)
        
        return MarketFeature(
            returns=returns, volume=volume,
            volatility=volatility, sma_ratio=sma_ratio
        )
    
    def create_sequences(self, features: MarketFeature) -> Tuple[np.ndarray, np.ndarray]:
        n = len(features.returns)
        if n <= self.lookback + self.prediction_horizon:
            return np.array([]), np.array([])
        
        X, y = [], []
        for i in range(self.lookback, n - self.prediction_horizon):
            seq = np.column_stack([
                features.returns[i-self.lookback:i],
                features.volume[i-self.lookback:i],
                features.volatility[i-self.lookback:i],
                features.sma_ratio[i-self.lookback:i]
            ])
            future_return = (features.returns[i+self.prediction_horizon] > 0).astype(float)
            X.append(seq)
            y.append(future_return)
        
        return np.array(X), np.array(y)
    
    def sigmoid(self, x):
        return 1 / (1 + np.exp(-np.clip(x, -500, 500)))
    
    def tanh(self, x):
        return np.tanh(x)
    
    def softmax(self, x):
        exp_x = np.exp(x - np.max(x, axis=-1, keepdims=True))
        return exp_x / np.sum(exp_x, axis=-1, keepdims=True)
    
    def lstm_forward(self, X, weights, bias, hidden_size):
        batch_size, seq_len, features = X.shape
        h = np.zeros((batch_size, hidden_size))
        C = np.zeros((batch_size, hidden_size))
        
        for t in range(seq_len):
            x_t = X[:, t, :]
            concat = np.concatenate([h, x_t], axis=1)
            
            wf = weights['wf']; wi = weights['wi']; wc = weights['wc']
            wo = weights['wo']; bf = bias['bf']; bi = bias['bi']
            bc = bias['bc']; bo = bias['bo']
            
            f = self.sigmoid(concat @ wf.T + bf)
            i = self.sigmoid(concat @ wi.T + bi)
            C_temp = self.tanh(concat @ wc.T + bc)
            o = self.sigmoid(concat @ wo.T + bo)
            
            C = f * C + i * C_temp
            h = o * self.tanh(C)
        
        return h, C
    
    def attention_forward(self, h_seq, attention_weights):
        v = attention_weights['v']
        Wa = attention_weights['Wa']
        
        e = np.tanh(h_seq @ Wa.T) @ v.T
        alpha = self.softmax(e)
        context = np.sum(alpha[:, :, np.newaxis] * h_seq, axis=1)
        return context, alpha
    
    def forward(self, X):
        h1, _ = self.lstm_forward(X, self.lstm1_weights, self.lstm1_bias, 64)
        context, _ = self.attention_forward(h1[:, np.newaxis, :], self.attention_weights)
        h2, _ = self.lstm_forward(context[:, np.newaxis, :], self.lstm2_weights, self.lstm2_bias, 32)
        
        h3 = np.maximum(0, h2 @ self.dense1_weights.T + self.dense1_bias)
        logits = h3 @ self.output_weights.T + self.output_bias
        probs = self.sigmoid(logits)
        return probs
    
    def initialize_weights(self, input_shape):
        feature_dim = input_shape[2]
        self.lstm1_weights = {
            'wf': np.random.randn(64, 64 + feature_dim) * 0.01,
            'wi': np.random.randn(64, 64 + feature_dim) * 0.01,
            'wc': np.random.randn(64, 64 + feature_dim) * 0.01,
            'wo': np.random.randn(64, 64 + feature_dim) * 0.01,
        }
        self.lstm1_bias = {k: np.zeros(64) for k in ['bf', 'bi', 'bc', 'bo']}
        
        self.attention_weights = {
            'v': np.random.randn(64) * 0.01,
            'Wa': np.random.randn(64, 64) * 0.01,
        }
        
        self.lstm2_weights = {
            k: np.random.randn(32, 32 + 64) * 0.01 
            for k in ['wf', 'wi', 'wc', 'wo']
        }
        self.lstm2_bias = {k: np.zeros(32) for k in ['bf', 'bi', 'bc', 'bo']}
        
        self.dense1_weights = np.random.randn(16, 32) * 0.01
        self.dense1_bias = np.zeros(16)
        self.output_weights = np.random.randn(1, 16) * 0.01
        self.output_bias = np.zeros(1)
    
    def fit(self, X, y, epochs=50, learning_rate=0.001):
        self.initialize_weights(X.shape)
        
        for epoch in range(epochs):
            probs = self.forward(X)
            error = y - probs
            
            if epoch % 10 == 0:
                pred = (probs > 0.5).astype(int)
                acc = np.mean(pred == y)
                loss = -np.mean(y * np.log(probs + 1e-10) + 
                              (1 - y) * np.log(1 - probs + 1e-10))
                print(f"Epoch {epoch}, Loss: {loss:.4f}, Accuracy: {acc:.2%}")
    
    def predict(self, X) -> List[TrendSignal]:
        probs = self.forward(X)
        signals = []
        
        for i, prob in enumerate(probs):
            direction = 'buy' if prob > 0.6 else ('sell' if prob < 0.4 else 'hold')
            confidence = max(prob, 1 - prob) if 0.4 < prob < 0.6 else 0.5
            signals.append(TrendSignal(
                direction=direction,
                confidence=confidence,
                predicted_return=prob - 0.5,
                holding_period=self.prediction_horizon
            ))
        return signals

# 示例运行
if __name__ == '__main__':
    np.random.seed(42)
    n = 500
    prices = 100 * np.exp(np.cumsum(np.random.randn(n) * 0.02))
    volumes = np.random.randint(1000000, 10000000, n)
    
    data = pd.DataFrame({
        'close': prices,
        'volume': volumes
    })
    
    predictor = MLTrendPredictor(lookback=60, prediction_horizon=5)
    features = predictor.compute_features(data)
    X, y = predictor.create_sequences(features)
    
    if len(X) > 0:
        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
        
        predictor.fit(X_train, y_train, epochs=50)
        signals = predictor.predict(X_test[:10])
        
        for i, sig in enumerate(signals):
            print(f"Sample {i}: {sig.direction}, confidence={sig.confidence:.2%}")
